### PR TITLE
Feature/authentication

### PR DESCRIPTION
**Pull request by toneo-**
In this PR:
- Can now log in and can choose Admin or User on the home page.
- The current username is stored in the session as "username". The current User is automatically pulled from the database and stored in the session as "user".
- Sessions last for 20 minutes as per assignment spec
- If you aren't logged in and you try to access /admin, /claim or /member you will be redirected to /auth.

Would appreciate it if you can check my unit tests pass on your machine (select all the *Test.java files, right click and run tests).

### CODE REVIEW

File: XYZDrivers/pom.xml [hunk 1]
```diff
@@ -47,8 +60,8 @@
                 <artifactId>maven-compiler-plugin</artifactId>
                 <version>3.1</version>
                 <configuration>
-                    <source>1.7</source>
-                    <target>1.7</target>
+                    <source>1.8</source>
```
Comment (bkneis) [inline XYZDrivers/pom.xml]: Why did you change the version of java? I think we were meant to use 1.7 and I'm not sure if the team have 1.8 installed on their machines

File: XYZDrivers/src/main/java/com/xyzdrivers/controllers/AuthController.java [hunk 2]
```diff
@@ -70,7 +63,47 @@ protected void doGet(HttpServletRequest request, HttpServletResponse response)
...
+        
+        String username = (String)request.getParameter("username");
+        String password = (String)request.getParameter("password");
+        String userType = (String)request.getParameter("userType");
+        
+        User loggedInUser = null;
+        HttpSession session = request.getSession(true);
+                
+        try {
+            if (userService.checkLoginDetails(username, password))
```
Comment (bkneis) [inline XYZDrivers/src/main/java/com/xyzdrivers/controllers/AuthController.java:73]: Please be consistent with the code formatting. Seems that { starts on a new line and sometimes doesn't

File: XYZDrivers/src/main/java/com/xyzdrivers/controllers/ClaimsController.java [hunk 3]
```diff
@@ -57,7 +58,10 @@ protected void doPost(HttpServletRequest request, HttpServletResponse response)
             String reason = request.getParameter("reason");
             String amount = request.getParameter("amount");
             
-            Claim claim = new Claim("null", date, reason, "NEW", Float.parseFloat(amount));
+            HttpSession session = request.getSession();
+            String username = (String)session.getAttribute("username");
```
Comment (bkneis) [inline XYZDrivers/src/main/java/com/xyzdrivers/controllers/ClaimsController.java]: Redundant cast?

File: XYZDrivers/src/main/java/com/xyzdrivers/filters/AuthorisationFilter.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        
+        chain.doFilter(request, response);
+        
+    }
+    
+    private String getLoggedInUsername(HttpSession session)
+    {
+        Object rawValue = session.getAttribute("username");
+        
+        return (rawValue == null) ? null : (String)rawValue;
```
Comment (bkneis) [inline XYZDrivers/src/main/java/com/xyzdrivers/filters/AuthorisationFilter.java]: No need for () using tenary op

File: XYZDrivers/src/main/java/com/xyzdrivers/repositories/UserRepo.java [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+        // TODO - specify primary key when calling on sqlService. SQLService is currently broken.
+        Stream<Object[]> allUsers;
+        try {
+            allUsers = sqlService.retrieve("users").stream();
+        } catch (SQLException ex) {
+            throw new RepositoryException("Retrieval failed.", ex);
+        }
+        
+        Object[] matches = allUsers
+                .filter(user -> ((String)user[0]).equals(id))
```
Comment (bkneis) [inline XYZDrivers/src/main/java/com/xyzdrivers/repositories/UserRepo.java]: Please do not use a different version of java than the spec. I personally do not have 1.8 ee on my machine

File: XYZDrivers/src/main/java/com/xyzdrivers/services/SQLService.java [hunk 6]
```diff
@@ -328,6 +328,48 @@ public void update(String table, String column, String primaryKeyColumn, Object
...
+        }
+                
+        if (columnValues.isEmpty()) {
+            return;
+        }
+        
+        String updateTemplate = "UPDATE ? SET ? = ? WHERE ? = ?";
+        
+        for (ColumnValuePair pair : columnValues)
+        {
```
Comment (bkneis) [inline XYZDrivers/src/main/java/com/xyzdrivers/services/SQLService.java:369]: same line for { please

File: XYZDrivers/src/test/java/com/xyzdrivers/tests/UserServiceTests.java [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+    public void Before()
+    {
+        userRepo = mock(UserRepo.class);
+        userService = new UserService(userRepo);
+    }
+    
+    @Test
+    public void checkLoginDetails_shouldThrowExceptionIfUserIdIsNull()
+    {
+        Executable nullTest = () -> {
```
Comment (bkneis) [inline XYZDrivers/src/test/java/com/xyzdrivers/tests/UserServiceTests.java]: 1.8 usage

File: XYZDrivers/src/main/java/com/xyzdrivers/services/UserService.java [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+        
+        if (userId == null) {
+            throw new IllegalArgumentException("userId cannot be null or blank.");
+        }
+        
+        if (password == null) {
+            throw new IllegalArgumentException("password cannot be null or blank.");
+        }
+        
+        User user = userRepo.get(userId);
```
Comment (bkneis) [inline XYZDrivers/src/main/java/com/xyzdrivers/services/UserService.java]: This would fail if someone initiates the service using a parameter less constructor. Please add a sanity check

**Review by bkneis — COMMENTED** [review]
Code looks great! Just a few comments before merging, mainly not using java 1.8

File: XYZDrivers/pom.xml (hunk 1 above)
Comment (lenfred123) [reply XYZDrivers/pom.xml]: I have Java 8 installed on my machine but can't speak for anyone else. I think we should stick with 1.7 as that's probably what the assignment spec asks for.

**Comment by lenfred123** [comment]
Ran the 22 tests and they all passed. Haven't done an explicit review as Bryan has already done one.

File: XYZDrivers/src/main/java/com/xyzdrivers/services/UserService.java (hunk 8 above)
Comment (toneo-) [reply XYZDrivers/src/main/java/com/xyzdrivers/services/UserService.java]: Fair enough

File: XYZDrivers/pom.xml (hunk 1 above)
Comment (toneo-) [reply XYZDrivers/pom.xml]: I did it to get lambdas, but since you don't have it I'll convert away from them. Also a good call just in case the demo computers don't have Java 8.

**Review by bkneis — APPROVED** [review]
Nice work!

File: XYZDrivers/src/main/java/com/xyzdrivers/controllers/ClaimsController.java (hunk 3 above)
Comment (toneo-) [reply XYZDrivers/src/main/java/com/xyzdrivers/controllers/ClaimsController.java]: Required, can't implicitly cast String to Object otherwise. Surprised me too.
